examples/transactions/transactional_producer.go

```diff
@@ -0,0 +1,103 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/twmb/franz-go/pkg/kgo"
+)
+
+func startProducing(ctx context.Context, kafkaBrokers string, topic string) {
+	producerId := strconv.FormatInt(int64(os.Getpid()), 10)
+	client, err := kgo.NewClient(
+		kgo.SeedBrokers(strings.Split(kafkaBrokers, ",")...),
+		kgo.TransactionalID(producerId),
+	)
+	if err != nil {
+		fmt.Printf("error initializing Kafka producer client: %v", err)
+		return
+	}
+
+	defer client.Close()
+
+	batch := 0
+	for {
+		if err := client.BeginTransaction(); err != nil {
+			fmt.Printf("error beginning transaction: %v\n", err)
+			break
+		}
+
+		// Write some messages in the transaction.
+		if err := produceRecords(ctx, client, topic, batch); err != nil {
+			fmt.Printf("error producing message: %v\n", err)
+			rollback(ctx, client)
+			continue
+		}
+
+		// Flush all of the buffered messages.
+		if err := client.Flush(ctx); err != nil {
+			if err != context.Canceled {
+				fmt.Printf("error flushing messages: %v\n", err)
+			}
+
+			rollback(ctx, client)
+			continue
+		}
+
+		// Attempt to commit the transaction and explicitly abort if it fails.
+		if err := client.EndTransaction(ctx, kgo.TryCommit); err != nil {
+			fmt.Printf("error committing transaction: %v\n", err)
+
+			if rollbackErr := client.AbortBufferedRecords(ctx); err != nil {
+				fmt.Printf("error rolling back transaction after commit failure: %v\n", rollbackErr)
+			}
+		}
+
+		batch += 1
+		time.Sleep(10 * time.Second)
+	}
+
+	fmt.Println("producer exited")
+}
+
+func produceRecords(ctx context.Context, client *kgo.Client, topic string, batch int) error {
+	errChan := make(chan error)
+
+	for i := 0; i < 10; i++ {
+		message := fmt.Sprintf("batch %d record %d\n", batch, i)
+		r := &kgo.Record{
+			Value: []byte(message),
+			Topic: topic,
+		}
+
+		err := client.Produce(ctx, r, func(r *kgo.Record, e error) {
+			fmt.Printf("produced message: %s", message)
+			errChan <- e
+		})
+		if err != nil {
+			return err
+		}
+
+		select {
```

Comment:
This select forces the loop to wait until a `ProduceRequest` is actually issued, meaning each produce is sequential and blocking. What you likely want to do is to produce all 10 records as fast as possible.

In this mini example, you may want to pass in the `errChan` to `produceRecords` and in `startProducing`, have a goroutine draining `errChan` and reacting to what it receives.

Reply:
Or is this meant to demonstrate that the consumer doesn't consume these messages until the transaction is done?

Reply:
That's exactly what I had in mind, I was trying to demonstrate that "batches" of messages are committed at once and the consumer will not read them until the batch is ready. However I'm happy to change this if you think it's clearer

Reply:
Am I right in thinking that it's actually unnecessary to check `ctx.Done` explicitly since `Produce` will return the context cancellation error if that occurs?

Reply:
`Produce` itself only returns an error in a very few select cases, so if you know what you're doing (not trying to produce when not in a transaction, not buffering max records and then canceling the context), you don't really _need_ to inspect the error.

But yes, you're right that it's not necessary to check `ctx.Done` here. The context is only used if the client is currently at the max buffered records, if it is, it waits for some records to finish or for the context to be canceled.

Reply:
> However I'm happy to change this if you think it's clearer
I think that since the point is to show the consumer doesn't consume when using ReadCommitted until the transaction is done, this is good. Maybe add some preceding documentation to explain that purpose?
```go
// We wait sequentially for every record to be produced,
// demonstrating that the consumer does not consume
// with ReadCommitted until the transaction is ended.
```
or something